third/nmh/man/mhshow.man
.\"
.\" %nmhwarning%
.\" $Id: mhshow.man,v 1.1.1.1 1999-02-07 18:14:21 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH MHSHOW %manext1% MH.6.8 [%nmhversion%]
.SH NAME
mhshow \- display MIME messages
.SH SYNOPSIS
.in +.5i
.ti -.5i
mhshow \%[+folder] \%[msgs] \%[\-file file]
.br
\%[\-part number]... \%[\-type content]...
.br
\%[\-serialonly] \%[\-noserialonly]
\%[\-pause] \%[\-nopause]
.br
\%[\-check] \%[\-nocheck]
\%[\-form formfile]
.br
\%[\-rcache policy] \%[\-wcache policy]
.br
\%[\-verbose] \%[\-noverbose]
\%[\-version] \%[\-help]
.in -.5i

.SH DESCRIPTION
The \fImhshow\fR command display contents of a MIME (multi-media)
message or collection of messages.

\fImhshow\fR manipulates multi-media messages as specified in
RFC\-2045 thru RFC\-2049.  Currently \fImhshow\fR only supports
encodings in message bodies, and does not support the encoding of
message headers as specified in RFC\-2047.

By default \fImhshow\fR will display all parts of a multipart
message.  By using the `\-part' and `\-type' switches, you may
limit the scope of \fImhshow\fR to particular subparts (of a
multipart content) and/or particular content types.

The option `\-file\ file' directs \fImhshow\fR to use the specified file as
the source message, rather than a message from a folder.  If you specify
this file as \*(lq-\*(rq, then \fImhshow\fR will accept the source message
on the standard input.  Note that the file, or input from standard input
should be a validly formatted message, just like any other \fInmh\fR
message.  It should \fBNOT\fR be in mail drop format (to convert a file in
mail drop format to a folder of \fInmh\fR messages, see \fIinc\fR\0(1)).

A part specification consists of a series of numbers separated by dots.
For example, in a multipart content containing three parts, these
would be named as 1, 2, and 3, respectively.  If part 2 was also a
multipart content containing two parts, these would be named as 2.1 and
2.2, respectively.  Note that the `\-part' switch is effective for only
messages containing a multipart content.  If a message has some other
kind of content, or if the part is itself another multipart content, the
`\-part' switch will not prevent the content from being acted upon.

A content specification consists of a content type and a subtype.
The initial list of \*(lqstandard\*(rq content types and subtypes can
be found in RFC\-2046.
.ne 18
A list of commonly used contents is briefly reproduced here:
.sp
.nf
.in +.5i
.ta \w'application  'u
Type	Subtypes
----	--------
text	plain, enriched
multipart	mixed, alternative, digest, parallel
message	rfc822, partial, external-body
application	octet-stream, postscript
image	jpeg, gif, png
audio	basic
video	mpeg
.re
.in -.5i
.fi
.sp
A legal MIME message must contain a subtype specification.
.PP
To specify a content, regardless of its subtype, just use the
name of the content, e.g., \*(lqaudio\*(rq.  To specify a specific
subtype, separate the two with a slash, e.g., \*(lqaudio/basic\*(rq.
Note that regardless of the values given to the `\-type' switch, a
multipart content (of any subtype listed above) is always acted upon.
Further note that if the `\-type' switch is used, and it is desirable to
act on a message/external-body content, then the `\-type' switch must
be used twice: once for message/external-body and once for the content
externally referenced.

.Uh "Unseen Sequence"

If the profile entry \*(lqUnseen\-Sequence\*(rq is present and
non\-empty, then \fImhshow\fR will remove each of the messages shown
from each sequence named by the profile entry.

.Uh "Checking the Contents"
The `\-check' switch tells \fImhshow\fR to check each content for an
integrity checksum.  If a content has such a checksum (specified as a
Content-MD5 header field), then \fImhshow\fR will attempt to verify the
integrity of the content.

.Uh "Showing the Contents"
The headers of each message are displayed with the \fImhlproc\fR
(usually \fImhl\fR), using the standard format file \fImhl.headers\fR.
You may specify an alternate format file with the `\-form formfile'
switch.  If the format file \fImhl.null\fR is specified, then the display
of the message headers is suppressed.

The method used to display the different contents in the messages bodies
will be determined by a \*(lqdisplay string\*(rq.  To find the display
string, \fImhshow\fR will first search your profile for an entry of the
form:
.sp
.in +.5i
mhshow-show-<type>/<subtype>
.in -.5i
.sp
to determine the display string.  If this isn't found, \fImhshow\fR
will search for an entry of the form:
.sp
.in +.5i
mhshow-show-<type>
.in -.5i
.sp
to determine the display string.

If a display string is found, any escapes (given below) will be expanded.
The result will be executed under \fB/bin/sh\fR, with the standard input
set to the content.
.ne 16
The display string may contain the following escapes:
.sp
.nf
.in +.5i
.ta \w'%F  'u
%a	Insert parameters from Content-Type field
%e	exclusive execution
%f	Insert filename containing content
%F	%e, %f, and stdin is terminal not content
%l	display listing prior to displaying content
%p	%l, and ask for confirmation
%s	Insert content subtype
%d	Insert content description
%%	Insert the character %
.re
.in -.5i
.fi
.sp
.ne 10
For those display strings containing the e- or F-escape, \fImhshow\fR will
execute at most one of these at any given time.  Although the F-escape
expands to be the filename containing the content, the e-escape has no
expansion as far as the shell is concerned.

When the p-escape prompts for confirmation, typing INTR (usually
control-C) will tell \fImhshow\fR not to display that content.
The p-escape can be disabled by specifying the switch `\-nopause'.
Further, when \fImhshow\fR is display a content, typing QUIT (usually
control-\\) will tell \fImhshow\fR to wrap things up immediately.

Note that if the content being displayed is multipart, but not one of
the subtypes listed above, then the f- and F-escapes expand to multiple
filenames, one for each subordinate content.  Further, stdin is not
redirected from the terminal to the content.

If a display string is not found, \fImhshow\fR has several default values:
.sp
.nf
.in +.5i
mhshow-show-text/plain: %pmoreproc '%F'
mhshow-show-message/rfc822: %pshow -file '%F'
.in -.5i
.fi
.sp
If a subtype of type text doesn't have a profile entry, it will be
treated as text/plain.

\fImhshow\fR has default methods for handling multipart messages of subtype
mixed, alternative, parallel, and digest.  Any unknown subtype of type
multipart (without a profile entry), will be treated as multipart/mixed.

If none of these apply, then \fImhshow\fR will check to see if the message
has an application/octet-stream content with parameter \*(lqtype=tar\*(rq.
If so, \fImhshow\fR will use an appropriate command.  If not, \fImhshow\fR
will complain.

.ne 10
Example entries might be:
.sp
.nf
.in +.5i
mhshow-show-audio/basic: raw2audio 2>/dev/null | play
mhshow-show-image: xv '%f'
mhshow-show-application/PostScript: lpr -Pps
.in -.5i
.fi
.sp
Note that when using the f- or F-escape, it's a good idea to use
single-quotes around the escape.  This prevents misinterpretation by
the shell of any funny characters that might be present in the filename.

Finally, \fImhshow\fR will process each message serially\0--\0it won't start
showing the next message until all the commands executed to display the
current message have terminated.  In the case of a multipart content
(of any subtype listed above), the content contains advice indicating if
the parts should be displayed serially or in parallel.  Because this may
cause confusion, particularly on uni-window displays, the `\-serialonly'
switch can be given to tell \fImhshow\fR to never display parts in parallel.

.Uh "Showing Alternate Character Sets"
Because a content of type text might be in a non-ASCII character
set, when \fImhshow\fR encounters a \*(lqcharset\*(rq parameter for
this content, it checks if your terminal can display this character
set natively.  \fIMhn\fR checks this by examining the the environment
variable MM_CHARSET.  If the value of this environment variable is equal
to the value of the charset parameter, then \fImhshow\fR assumes it can
display this content without any additional setup.  If this environment
variable is not set, \fImhshow\fR will assume a value of \*(lqUS-ASCII\*(rq.
If the character set cannot be displayed natively, then \fImhshow\fR will
look for an entry of the form:
.sp
.in +.5i
mhshow-charset-<charset>
.in -.5i
.sp
which should contain a command creating an environment to render
the character set.  This command string should containing a single
\*(lq%s\*(rq, which will be filled-in with the command to display the
content.

Example entries might be:
.sp
.in +.5i
mhshow-charset-iso-8859-1: xterm -fn '-*-*-medium-r-normal-*-*-120-*-*-c-*-iso8859-*' -e %s
.in -.5i
or
.in +.5i
mhshow-charset-iso-8859-1: '%s'
.in -.5i
.sp
The first example tells \fImhshow\fR to start \fIxterm\fR and load the
appropriate character set for that message content.  The second example
tells \fImhshow\fR that your pager (or other program handling that content
type) can handle that character set, and that no special processing is
needed beforehand.
.sp
Note that many pagers strip off the high-order bit or have problems
displaying text with the high-order bit set.  However, the pager
\fIless\fR has support for single-octet character sets.  The source
to \fIless\fR is available on many ftp sites carrying free software.
In order to view messages sent in the ISO-8859-1 character set using
\fIless\fR,
.ne 9
put these lines in your \&.login file:
.sp
.nf
.in +.5i
setenv LESSCHARSET latin1
setenv LESS "-f"
.in -.5i
.fi
.sp
The first line tells \fIless\fR to use the ISO-8859-1 definition for
determining whether a character is \*(lqnormal\*(rq, \*(lqcontrol\*(lq,
or \*(lqbinary\*(rq.  The second line tells \fIless\fR not to warn you
if it encounters a file that has non-ASCII characters.  Then, simply
set the \fBmoreproc\fR profile entry to \fIless\fR, and it will get
called automatically.  (To handle other single-octet character sets,
look at the \fIless\fR\0(1) manual entry for information about the
\fBLESSCHARDEF\fR environment variable.)

.Uh "Messages of Type message/partial"
\fImhshow\fR cannot directly display messages of type partial.
You must reassemble them first into a normal message using
\fImhstore\fR.  Check the man page for \fImhstore\fR for details.

.Uh "External Access"
For contents of type message/external-body,
.ne 12
\fImhshow\fR supports these access-types:
.sp
.nf
.in +.5i
afs
anon-ftp
ftp
local-file
mail-server
.in -.5i
.fi
.sp
For the \*(lqanon-ftp\*(rq and \*(lqftp\*(rq access types,
\fImhshow\fR will look for the \fBnmh-access-ftp\fR
profile entry,
.ne 6
e.g.,
.sp
.in +.5i
nmh-access-ftp: myftp.sh
.in -.5i
.sp
to determine the pathname of a program to perform the FTP retrieval.
.ne 14
This program is invoked with these arguments:
.sp
.nf
.in +.5i
domain name of FTP-site
username
password
remote directory
remote filename
local filename
\*(lqascii\*(rq or \*(lqbinary\*(rq
.in -.5i
.fi
.sp
The program should terminate with an exit status of zero if the
retrieval is successful, and a non-zero exit status otherwise.

If this entry is not provided, then \fImhshow\fR will use a simple
built-in FTP client to perform the retrieval.

.Uh "The Content Cache"
When \fImhshow\fR encounters an external content containing a
\*(lqContent-ID:\*(rq field, and if the content allows caching, then
depending on the caching behavior of \fImhshow\fR, the content might be
read from or written to a cache.

The caching behavior of \fImhshow\fR is controlled with the `\-rcache'
and `\-wcache' switches, which define the policy for reading from,
and writing to, the cache, respectively.  One of four policies may be
specified: \*(lqpublic\*(rq, indicating that \fImhshow\fR should make use
of a publically-accessible content cache; \*(lqprivate\*(rq, indicating
that \fImhshow\fR should make use of the user's private content cache;
\*(lqnever\*(rq, indicating that \fImhshow\fR should never make use of
caching; and, \*(lqask\*(rq, indicating that \fImhshow\fR should ask
the user.

There are two directories where contents may be cached: the profile entry
\fBnmh-cache\fR names a directory containing world-readable contents, and,
the profile entry \fBnmh-private-cache\fR names a directory containing
private contents.  The former should be an absolute (rooted) directory
name.
.ne 6
For example,
.sp
.in +.5i
nmh-cache: /tmp
.in -.5i
.sp
might be used if you didn't care that the cache got wiped after each
reboot of the system.  The latter is interpreted relative to the user's
nmh directory, if not rooted,
.ne 6
e.g.,
.sp
.in +.5i
nmh-private-cache: .cache
.in -.5i
.sp
(which is the default value).

.Uh "User Environment"
Because the display environment in which \fImhshow\fR operates may vary for
different machines, \fImhshow\fR will look for the environment variable
\fB$MHSHOW\fR.  If present, this specifies the name of an additional
user profile which should be read.  Hence, when a user logs in on a
particular display device, this environment variable should be set to
refer to a file containing definitions useful for the given display device.
Normally, only entries that deal with the methods to display different
content type and subtypes
.sp
.in +.5i
mhshow-show-<type>/<subtype>
.br
mhshow-show-<type>
.in -.5i
.sp
need be present in this additional profile.
Finally,
\fImhshow\fR will attempt to consult one other additional user profile,
.ne 6
e.g.,
.sp
.in +.5i
%etcdir%/mhn.defaults
.in -.5i
.sp
which is created automatically during nmh installation.
.Fi
^$HOME/\&.mh\(ruprofile~^The user profile
^$MHSHOW~^Additional profile entries
^%etcdir%/mhn.defaults~^System default MIME profile entries
^%etcdir%/mhl.headers~^The headers template
.Pr
^Path:~^To determine the user's nmh directory
.Ps
^Current\-Folder:~^To find the default current folder
.Ps
^Unseen\-Sequence:~^To name sequences denoting unseen messages
.Ps
^mhlproc:~^Default program to display message headers
.Ps
^nmh-access-ftp:~^Program to retrieve contents via FTP
.Ps
^nmh-cache~^Public directory to store cached external contents
.Ps
^nmh-private-cache~^Personal directory to store cached external contents
.Ps
^mhshow-charset-<charset>~^Template for environment to render character sets
.Ps
^mhshow-show-<type>*~^Template for displaying contents
.Ps
^moreproc:~^Default program to display text/plain content
.Sa
mhbuild(1), mhl(1), mhlist(1), mhstore(1), sendfiles(1)
.br
RFC\-934:
.br
   \fIProposed Standard for Message Encapsulation\fR,
.br
RFC\-2045:
.br
   \fIMultipurpose Internet Mail Extensions (MIME) Part One:
.br
   Format of Internet Message Bodies\fR,
.br
RFC\-2046:
.br
   \fIMultipurpose Internet Mail Extensions (MIME) Part Two:
.br
   Media Types\fR,
.br
RFC\-2047:
.br
   \fIMultipurpose Internet Mail Extensions (MIME) Part Three:
.br
   Message Header Extensions for Non-ASCII Text\fR,
.br
RFC\-2048:
.br
   \fIMultipurpose Internet Mail Extensions (MIME) Part Four:
.br
   Registration Procedures\fR,
.br
RFC\-2049:
.br
   \fIMultipurpose Internet Mail Extensions (MIME) Part Five:
.br
   Conformance Criteria and Examples\fR.
.De
`+folder' defaults to the current folder
.Ds
`msgs' defaults to cur
.Ds
`\-nocheck'
.Ds
`\-form mhl.headers'
.Ds
`\-pause'
.Ds
`\-rcache ask'
.Ds
`\-realsize'
.Ds
`\-noserialonly'
.Ds
`\-noverbose'
.Ds
`\-wcache ask'
.Co
If a folder is given, it will become the current folder.  The last
message selected will become the current message.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/mhshow.c

/*
 * mhshow.c -- display the contents of MIME messages
 *
 * $Id: mhshow.c,v 1.1.1.1 1999-02-07 18:14:15 danw Exp $
 */

#include <h/mh.h>
#include <fcntl.h>
#include <h/signals.h>
#include <h/md5.h>
#include <errno.h>
#include <signal.h>
#include <zotnet/mts/mts.h>
#include <zotnet/tws/tws.h>
#include <h/mime.h>
#include <h/mhparse.h>
#include <h/mhcachesbr.h>

#ifdef HAVE_SYS_WAIT_H
# include <sys/wait.h>
#endif

/*
 * We allocate space for message names (msgs array)
 * this number of elements at a time.
 */
#define MAXMSGS  256


static struct swit switches[] = {
#define	CHECKSW                 0
    { "check", 0 },
#define	NCHECKSW                1
    { "nocheck", 0 },
#define	PAUSESW                 2
    { "pause", 0 },
#define	NPAUSESW                3
    { "nopause", 0 },
#define	SERIALSW                4
    { "serialonly", 0 },
#define	NSERIALSW               5
    { "noserialonly", 0 },
#define	VERBSW                  6
    { "verbose", 0 },
#define	NVERBSW                 7
    { "noverbose", 0 },
#define	FILESW                  8	/* interface from show */
    { "file file", 0 },
#define	FORMSW                  9
    { "form formfile", 0 },
#define	PARTSW                 10
    { "part number", 0 },
#define	TYPESW                 11
    { "type content", 0 },
#define	RCACHESW               12
    { "rcache policy", 0 },
#define	WCACHESW               13
    { "wcache policy", 0 },
#define VERSIONSW              14
    { "version", 0 },
#define	HELPSW                 15
    { "help", 4 },

/*
 * switches for moreproc/mhlproc
 */
#define	PROGSW                 16
    { "moreproc program", -4 },
#define	NPROGSW                17
    { "nomoreproc", -3 },
#define	LENSW                  18
    { "length lines", -4 },
#define	WIDTHSW                19
    { "width columns", -4 },

/*
 * switches for debugging
 */
#define	DEBUGSW                20
    { "debug", -5 },
    { NULL, 0 }
};


extern int errno;

/* mhparse.c */
extern int checksw;
extern char *tmp;	/* directory to place temp files */

/* mhcachesbr.c */
extern int rcachesw;
extern int wcachesw;
extern char *cache_public;
extern char *cache_private;

/* mhshowsbr.c */
extern int pausesw;
extern int serialsw;
extern char *progsw;
extern int nolist;
extern int nomore;	/* flags for moreproc/header display */
extern char *formsw;

/* mhmisc.c */
extern int npart;
extern int ntype;
extern char *parts[NPARTS + 1];
extern char *types[NTYPES + 1];
extern int userrs;

int debugsw = 0;
int verbosw = 0;

/* The list of top-level contents to display */
CT *cts = NULL;

#define	quitser	pipeser

/* mhparse.c */
CT parse_mime (char *);

/* mhmisc.c */
int part_ok (CT, int);
int type_ok (CT, int);
void set_endian (void);
void flush_errors (void);

/* mhshowsbr.c */
void show_all_messages (CT *);

/* mhfree.c */
void free_content (CT);

/*
 * static prototypes
 */
static RETSIGTYPE pipeser (int);


int
main (int argc, char **argv)
{
    int nummsgs, maxmsgs, msgnum, *icachesw;
    char *cp, *file = NULL, *folder = NULL;
    char *maildir, buf[100], **argp;
    char **arguments, **msgs;
    struct msgs *mp = NULL;
    CT ct, *ctp;
    FILE *fp;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /*
     * Allocate the initial space to record message
     * names, ranges, and sequences.
     */
    nummsgs = 0;
    maxmsgs = MAXMSGS;
    if (!(msgs = (char **) malloc ((size_t) (maxmsgs * sizeof(*msgs)))))
	adios (NULL, "unable to allocate storage");

    /*
     * Parse arguments
     */
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
	    case AMBIGSW: 
		ambigsw (cp, switches);
		done (1);
	    case UNKWNSW: 
		adios (NULL, "-%s unknown", cp);

	    case HELPSW: 
		snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			invo_name);
		print_help (buf, switches, 1);
		done (1);
	    case VERSIONSW:
		print_version(invo_name);
		done (1);

	    case RCACHESW:
		icachesw = &rcachesw;
		goto do_cache;
	    case WCACHESW:
		icachesw = &wcachesw;
do_cache:
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		switch (*icachesw = smatch (cp, caches)) {
		case AMBIGSW:
		    ambigsw (cp, caches);
		    done (1);
		case UNKWNSW:
		    adios (NULL, "%s unknown", cp);
		default:
		    break;
		}
		continue;

	    case CHECKSW:
		checksw++;
		continue;
	    case NCHECKSW:
		checksw = 0;
		continue;

	    case PAUSESW:
		pausesw = 1;
		continue;
	    case NPAUSESW:
		pausesw = 0;
		continue;

	    case SERIALSW:
		serialsw = 1;
		continue;
	    case NSERIALSW:
		serialsw = 0;
		continue;

	    case PARTSW:
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		if (npart >= NPARTS)
		    adios (NULL, "too many parts (starting with %s), %d max",
			   cp, NPARTS);
		parts[npart++] = cp;
		continue;

	    case TYPESW:
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		if (ntype >= NTYPES)
		    adios (NULL, "too many types (starting with %s), %d max",
			   cp, NTYPES);
		types[ntype++] = cp;
		continue;

	    case FILESW:
		if (!(cp = *argp++) || (*cp == '-' && cp[1]))
		    adios (NULL, "missing argument to %s", argp[-2]);
		file = *cp == '-' ? cp : path (cp, TFILE);
		continue;

	    case FORMSW:
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		if (formsw)
		    free (formsw);
		formsw = getcpy (etcpath (cp));
		continue;

	    /*
	     * Switches for moreproc/mhlproc
	     */
	    case PROGSW:
		if (!(progsw = *argp++) || *progsw == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		continue;
	    case NPROGSW:
		nomore++;
		continue;

	    case LENSW:
	    case WIDTHSW:
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);
		continue;

	    case VERBSW: 
		verbosw = 1;
		continue;
	    case NVERBSW: 
		verbosw = 0;
		continue;
	    case DEBUGSW:
		debugsw = 1;
		continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = path (cp + 1, *cp == '+' ? TFOLDER : TSUBCWF);
	} else {
	    /*
	     * Check if we need to allocate more space
	     * for message names/ranges/sequences.
	     */
	    if (nummsgs >= maxmsgs) {
		maxmsgs += MAXMSGS;
		if (!(msgs = (char **) realloc (msgs,
			(size_t) (maxmsgs * sizeof(*msgs)))))
		    adios (NULL, "unable to reallocate msgs storage");
	    }
	    msgs[nummsgs++] = cp;
	}
    }

    /* null terminate the list of acceptable parts/types */
    parts[npart] = NULL;
    types[ntype] = NULL;

    set_endian ();

    if ((cp = getenv ("MM_NOASK")) && !strcmp (cp, "1")) {
	nolist  = 1;
	pausesw = 0;
    }

    /*
     * Check if we've specified an additional profile
     */
    if ((cp = getenv ("MHSHOW"))) {
	if ((fp = fopen (cp, "r"))) {
	    readconfig ((struct node **) 0, fp, cp, 0);
	    fclose (fp);
	} else {
	    admonish ("", "unable to read $MHSHOW profile (%s)", cp);
	}
    }

    /*
     * Read the standard profile setup
     */
    if ((fp = fopen (cp = etcpath ("mhn.defaults"), "r"))) {
	readconfig ((struct node **) 0, fp, cp, 0);
	fclose (fp);
    }

    /* Check for public cache location */
    if ((cache_public = context_find (nmhcache)) && *cache_public != '/')
	cache_public = NULL;

    /* Check for private cache location */
    if (!(cache_private = context_find (nmhprivcache)))
	cache_private = ".cache";
    cache_private = getcpy (m_maildir (cache_private));

    /*
     * Check for storage directory.  If specified,
     * then store temporary files there.  Else we
     * store them in standard nmh directory.
     */
    if ((cp = context_find (nmhstorage)) && *cp)
	tmp = concat (cp, "/", invo_name, NULL);
    else
	tmp = add (m_maildir (invo_name), NULL);

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    if (file && nummsgs)
	adios (NULL, "cannot specify msg and file at same time!");

    /*
     * check if message is coming from file
     */
    if (file) {
	if (!(cts = (CT *) calloc ((size_t) 2, sizeof(*cts))))
	    adios (NULL, "out of memory");
	ctp = cts;

	if ((ct = parse_mime (file)));
	    *ctp++ = ct;
    } else {
	/*
	 * message(s) are coming from a folder
	 */
	if (!nummsgs)
	    msgs[nummsgs++] = "cur";
	if (!folder)
	    folder = getfolder (1);
	maildir = m_maildir (folder);

	if (chdir (maildir) == NOTOK)
	    adios (maildir, "unable to change directory to");

	/* read folder and create message structure */
	if (!(mp = folder_read (folder)))
	    adios (NULL, "unable to read folder %s", folder);

	/* check for empty folder */
	if (mp->nummsg == 0)
	    adios (NULL, "no messages in %s", folder);

	/* parse all the message ranges/sequences and set SELECTED */
	for (msgnum = 0; msgnum < nummsgs; msgnum++)
	    if (!m_convert (mp, msgs[msgnum]))
		done (1);

	/*
	 * Set the SELECT_UNSEEN bit for all the SELECTED messages,
	 * since we will use that as a tag to know which messages
	 * to remove from the "unseen" sequence.
	 */
	for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++)
	    if (is_selected(mp, msgnum))
		set_unseen (mp, msgnum);

	seq_setprev (mp);	/* set the Previous-Sequence */
	seq_setunseen (mp, 1);	/* unset the Unseen-Sequence */

	if (!(cts = (CT *) calloc ((size_t) (mp->numsel + 1), sizeof(*cts))))
	    adios (NULL, "out of memory");
	ctp = cts;

	/*
	 * Parse all the SELECTED messages.
	 */
	for (msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
	    if (is_selected(mp, msgnum)) {
		char *msgnam;

		msgnam = m_name (msgnum);
		if ((ct = parse_mime (msgnam)))
		    *ctp++ = ct;
	    }
	}
    }

    if (!*cts)
	done (1);

    userrs = 1;
    SIGNAL (SIGQUIT, quitser);
    SIGNAL (SIGPIPE, pipeser);

    /*
     * Get the associated umask for the relevant contents.
     */
    for (ctp = cts; *ctp; ctp++) {
	struct stat st;

	ct = *ctp;
	if (type_ok (ct, 1) && !ct->c_umask) {
	    if (stat (ct->c_file, &st) != NOTOK)
		ct->c_umask = ~(st.st_mode & 0777);
	    else
		ct->c_umask = ~m_gmprot();
	}
    }

    /*
     * Show the message content
     */
    show_all_messages (cts);

    /* Now free all the structures for the content */
    for (ctp = cts; *ctp; ctp++)
	free_content (*ctp);

    free ((char *) cts);
    cts = NULL;

    /* If reading from a folder, do some updating */
    if (mp) {
	context_replace (pfolder, folder);/* update current folder  */
	seq_setcur (mp, mp->hghsel);	  /* update current message */
	seq_save (mp);			  /* synchronize sequences  */
	context_save ();		  /* save the context file  */
    }

    done (0);
    /* NOTREACHED */
}


static RETSIGTYPE
pipeser (int i)
{
    if (i == SIGQUIT) {
	unlink ("core");
	fflush (stdout);
	fprintf (stderr, "\n");
	fflush (stderr);
    }

    done (1);
    /* NOTREACHED */
}


void
done (int status)
{
    CT *ctp;

    if ((ctp = cts))
	for (; *ctp; ctp++)
	    free_content (*ctp);

    exit (status);
}
